engine/crates/tracing/src/tower.rs

```diff
@@ -1,67 +1,175 @@
+    fn layer(&self, inner: S) -> Self::Service {
+        TracingService {
+            inner,
+            metrics: self.metrics.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct TracingService<S> {
+    inner: S,
+    metrics: RequestMetrics,
+}
+
+impl<S> TracingService<S> {
+    #[cfg(not(feature = "lambda"))]
+    fn make_span<B: Body>(&mut self, request: &Request<B>) -> Span {
+        HttpRequestSpan::from_http(request).into_span()
+    }
+
+    #[cfg(feature = "lambda")]
+    fn make_span<B: Body>(&self, request: &Request<B>) -> Span {
+        use opentelemetry::Context;
+        use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+        let parent_ctx = opentelemetry::global::get_text_map_propagator(|propagator| {
+            propagator.extract_with_context(&Context::current(), &HeaderExtractor(request.headers()))
+        });
+
+        let span = HttpRequestSpan::from_http(request).into_span();
+        span.set_parent(parent_ctx);
+
+        span
+    }
+}
+
+// From opentelemetry-http which still uses http 0.X as of 2024/05/17
+#[cfg_attr(not(feature = "lambda"), allow(unused))]
+struct HeaderExtractor<'a>(pub &'a http::HeaderMap);
+
+impl<'a> Extractor for HeaderExtractor<'a> {
+    /// Get a value for a key from the HeaderMap.  If the value is not valid ASCII, returns None.
+    fn get(&self, key: &str) -> Option<&str> {
+        self.0.get(key).and_then(|value| value.to_str().ok())
+    }
+
+    /// Collect all the keys from the HeaderMap.
+    fn keys(&self) -> Vec<&str> {
+        self.0.keys().map(|value| value.as_str()).collect::<Vec<_>>()
+    }
+}
+
+// Inspired by the tower-http Tracing layer which works nicely for spans but not as well for
+// metrics as we need access to both request headers (client) and response ones.
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for TracingService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    S::Error: std::fmt::Display + 'static,
+    ReqBody: Body,
+    ResBody: Body,
+{
+    type Response = http::Response<ResBody>;
+
+    type Error = S::Error;
+
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+        let client = Client::extract_from(req.headers());
+        let metrics = self.metrics.clone();
+        let span = self.make_span(&req);
+        ResponseFuture {
+            inner: self.inner.call(req),
+            metrics,
+            span,
+            start,
+            client,
+        }
+    }
+}
+
+pin_project! {
+    pub struct ResponseFuture<F> {
+        #[pin]
+        inner: F,
+        metrics: RequestMetrics,
+        span: Span,
+        start: Instant,
+        client: Option<Client>,
+    }
+}
+
+impl<F, ResBody, E> Future for ResponseFuture<F>
```

Comment:
Why are we implementing this directly?

Reply:
hum fair point, could have just returned a `BoxFuture` actually. Thought I could specify the future type in the trait initially but couldn't without creating a struct like `tower-http` did and went that way instead of falling back to a box